file_formats/cmt_format.ml

```diff
@@ -468,7 +468,12 @@ let save_cmt target binary_annots initial_env cmi shape =
            | None -> None
            | Some cmi -> Some (output_cmi temp_file_name oc cmi)
          in
-         let sourcefile = Unit_info.Artifact.source_file target in
+         (* We use the raw_source_file because the original_source_file may not
+            exist (or may have changed), so computing the digest may fail or
+            produce inconsistent results. Merlin expects the cmt_sourcefile to
+            be the file we computed the digest of, which is why we use the
+            raw_source_file for that as well. *)
```

Comment:
Another way to see this is that `cmt` files are metadata for a compiled file (with a typedtree that can contain objects that doesn't exist in the original source file), thus we should associate the `cmt` with the `raw` file rather than the original source file.

Reply:
Good point, I've added this to the comment.